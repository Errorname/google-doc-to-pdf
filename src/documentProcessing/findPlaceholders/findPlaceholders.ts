import { Placeholder } from '../types'
import { GDoc } from '../gdocTypes'

const findPlaceholders = (doc: GDoc): Placeholder[] => {
  const placeholders: Placeholder[] = []

  doc.body.content.map(c => {
    if (c.paragraph) {
      c.paragraph.elements.map(e => {
        if (e.textRun) {
          let textRun = e.textRun.content
          const matches = textRun.match(/{{([^}]*)}}/gi) || []
          for (let match of matches) {
            const start = e.startIndex + textRun.indexOf(match)
            textRun = textRun.replace(match, ''.padStart(match.length))
            placeholders.push({
              raw: match,
              position: {
                start,
                end: start + match.length
              }
            })
          }
        }
      })
    }
  })

  return placeholders
}

export default findPlaceholders
