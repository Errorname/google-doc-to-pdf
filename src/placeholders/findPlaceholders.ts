import { GDoc, Placeholder, StructuralElement } from '../types'

const findPlaceholders = (doc: GDoc): Placeholder[] => {
  const placeholders: Placeholder[] = []

  // Body
  placeholders.push(...findInContent(doc.body.content))

  // Headers
  for (const header of Object.values(doc.headers || {})) {
    placeholders.push(...findInContent(header.content, header.headerId))
  }

  // Footers
  for (const footer of Object.values(doc.footers || {})) {
    placeholders.push(...findInContent(footer.content, footer.footerId))
  }

  return placeholders
}

const findInContent = (content: StructuralElement[], segmentId?: string): Placeholder[] => {
  const placeholders: Placeholder[] = []

  for (const element of content) {
    // Paragraph
    if (element.paragraph) {
      for (const paragraphElement of element.paragraph.elements) {
        if (paragraphElement.textRun) {
          const textRun = paragraphElement.textRun.content
          const matches = textRun.match(/{{([^}]+)}}/gi) || []
          for (const match of matches) {
            const start = (paragraphElement.startIndex || 1) + textRun.indexOf(match)

            placeholders.push({
              raw: match,
              position: {
                start,
                end: start + match.length,
              },
              segmentId,
            })
          }
        }
      }
    }
    // Table
    else if (element.table) {
      element.table.tableRows.forEach((row) => {
        row.tableCells.forEach((cell) => {
          placeholders.push(...findInContent(cell.content, segmentId))
        })
      })
    }
  }

  return placeholders
}

export default findPlaceholders
