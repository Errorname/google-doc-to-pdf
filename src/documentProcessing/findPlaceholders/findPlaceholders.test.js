import findPlaceholders from './findPlaceholders'

describe('findPlaceholders', () => {
  test('empty document', () => {
    const document = {
      body: {
        content: []
      }
    }
    const placeholders = findPlaceholders(document)

    expect(placeholders).toEqual([])
  })

  test('document with unknown content', () => {
    const document = {
      body: {
        content: [{}, {}]
      }
    }
    const placeholders = findPlaceholders(document)

    expect(placeholders).toEqual([])
  })

  test('single paragraph with no placeholders', () => {
    const document = {
      body: {
        content: [
          {
            paragraph: {
              elements: [{ textRun: { content: 'Hello world!' } }]
            }
          }
        ]
      }
    }
    const placeholders = findPlaceholders(document)

    expect(placeholders).toEqual([])
  })

  test('multiple paragraphs with multiple placeholders', () => {
    const document = {
      body: {
        content: [
          {
            paragraph: {
              elements: [
                { textRun: { content: 'Hello {{ user.name }}!' }, startIndex: 0, endIndex: 21 }
              ]
            }
          },
          {
            paragraph: {
              elements: [
                {
                  textRun: { content: 'Have you seen {{movie.title | uppercase}}?' },
                  startIndex: 23,
                  endIndex: 64
                },
                {
                  textRun: {
                    content: 'In my opinion, I would rate it {{ movie.rating }} out of 5',
                    startIndex: 66,
                    endIndex: 123
                  }
                }
              ]
            }
          }
        ]
      }
    }
    const placeholders = findPlaceholders(document)

    expect(placeholders).toMatchSnapshot()
  })
})
