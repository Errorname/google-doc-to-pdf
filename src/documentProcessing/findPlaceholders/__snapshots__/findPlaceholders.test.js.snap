// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findPlaceholders multiple paragraphs with multiple placeholders 1`] = `
Array [
  Object {
    "position": Object {
      "end": 21,
      "start": 6,
    },
    "raw": "{{ user.name }}",
  },
  Object {
    "position": Object {
      "end": 64,
      "start": 37,
    },
    "raw": "{{movie.title | uppercase}}",
  },
  Object {
    "position": Object {
      "end": NaN,
      "start": NaN,
    },
    "raw": "{{ movie.rating }}",
  },
]
`;
